azkaban-executor/azkaban-emr-jobtype/src/main/java/uk/gov/dwp/dataworks/azkaban/services/CompositeService.java

```diff
@@ -0,0 +1,99 @@
+
+public class CompositeService extends DelegateService {
+
+
+    public CompositeService(DependencyService dependencyService,
+            LaunchInvocationService launchInvocationService, EmrProgressService emrProgressService,
+            NotificationService notificationService, StatusService statusService, AmazonElasticMapReduce emr) {
+        this.dependencyService = dependencyService;
+        this.launchInvocationService = launchInvocationService;
+        this.emrProgressService = emrProgressService;
+        this.notificationService = notificationService;
+        this.statusService = statusService;
+        this.emr = emr;
+    }
+
+    public boolean launchClusterAndWaitForStepCompletion(String dependency) {
+        try {
+            this.notificationService.notifyStarted();
+            boolean succeeded = dependencyMetadata(dependency)
+                    .map(this::registerDependenciesCompleted)
+                    .filter(x -> proceed.get())
+                    .flatMap(launchInvocationService::invokeEmrLauncher)
+                    .filter(InvocationResult::wasSuccessful)
+                    .map(InvocationResult::getClusterId)
+                    .map(this::setGetClusterId)
+                    .map(this::registerClusterId)
+                    .filter(x -> proceed.get())
+                    .map(emrProgressService::observeEmr)
+                    .orElse(false);
+
+            if (succeeded) {
+                this.statusService.registerSuccess();
+                this.notificationService.notifySucceeded();
+                return true;
+            } else {
+                this.statusService.registerFailure();
+                this.notificationService.notifyFailed();
+                return false;
+            }
+        } catch (Exception e) {
+            error("Job failed", e);
+            this.statusService.registerFailure();
+            this.notificationService.notifyFailed();
+            return false;
+        }
+    }
+
+    private String registerClusterId(String x) {
+        this.statusService.registerClusterId(x);
+        return x;
+    }
+
+    private InvocationPayload registerDependenciesCompleted(InvocationPayload payload) {
+        this.statusService.registerDependenciesCompleted(payload);
+        return payload;
+    }
+
+    @Override
+    public void cancel() {
+        super.cancel();
+        dependencyService.cancel();
+        launchInvocationService.cancel();
+        emrProgressService.cancel();
+        clusterId.get().ifPresent(id -> EmrUtility.cancelSteps(emr, id));
+    }
+
+    private Optional<InvocationPayload> dependencyMetadata(String dependencies) {
+        return completedDependency(dependencies)
+                .map(InvocationPayload::from);
+    }
+
+    private Optional<Map<String, AttributeValue>> completedDependency(String dependency) {
+        return dependencyService.successfulDependency(dependency);
+    }
+
+    private String setGetClusterId(String id) {
```

Comment:
badly named private method.

Reply:
It sets the cluster id then returns it. What would you prefer?